File: packages/shared/src/rarity.ts
// Rarity tiers are fixed at launch and mirrored in apps/api/cards/rarity.py.
// Keep both files in sync; the API test suite compares them against this file.
//
// Caps, odds, and recycle values can be tuned through playtesting.

export const RARITIES = ['common', 'uncommon', 'rare', 'epic', 'legendary'] as const;
export type Rarity = (typeof RARITIES)[number];

export const RARITY_LABELS: Record<Rarity, string> = {
  common: 'Common',
  uncommon: 'Uncommon',
  rare: 'Rare',
  epic: 'Epic',
  legendary: 'Legendary',
};

// Maximum share of a published set each tier may occupy. Common has no cap
// but the set must be at least COMMON_MIN_SHARE common so packs stay open-able.
export const RARITY_MAX_SHARE: Record<Exclude<Rarity, 'common'>, number> = {
  uncommon: 0.3,
  rare: 0.2,
  epic: 0.1,
  legendary: 0.05,
};
export const COMMON_MIN_SHARE = 0.4;

// Per-card pull probability inside a pack. Must sum to 1.
export const PULL_ODDS: Record<Rarity, number> = {
  common: 0.6,
  uncommon: 0.25,
  rare: 0.1,
  epic: 0.04,
  legendary: 0.01,
};

// Set-specific points granted when a duplicate of this rarity is recycled.
export const RECYCLE_VALUE: Record<Rarity, number> = {
  common: 1,
  uncommon: 3,
  rare: 10,
  epic: 30,
  legendary: 100,
};

export function isRarity(value: unknown): value is Rarity {
  return typeof value === 'string' && (RARITIES as readonly string[]).includes(value);
}

File: apps/api/cards/rarity.py
"""Rarity tiers and the rules attached to them.

Mirrors packages/shared/src/rarity.ts. A test in cards/tests/test_rarity.py
reads the TypeScript file and fails if the two drift apart.

Caps, odds, and recycle values can be tuned through playtesting.
"""

from collections import Counter

COMMON = "common"
UNCOMMON = "uncommon"
RARE = "rare"
EPIC = "epic"
LEGENDARY = "legendary"

RARITIES = [COMMON, UNCOMMON, RARE, EPIC, LEGENDARY]
RARITY_CHOICES = [(r, r.capitalize()) for r in RARITIES]

# Maximum share of a published set each tier may occupy.
RARITY_MAX_SHARE = {UNCOMMON: 0.30, RARE: 0.20, EPIC: 0.10, LEGENDARY: 0.05}
COMMON_MIN_SHARE = 0.40

PULL_ODDS = {COMMON: 0.60, UNCOMMON: 0.25, RARE: 0.10, EPIC: 0.04, LEGENDARY: 0.01}
RECYCLE_VALUE = {COMMON: 1, UNCOMMON: 3, RARE: 10, EPIC: 30, LEGENDARY: 100}

# Smallest set that can be published.
MIN_CARDS_TO_PUBLISH = 5


def rarity_problems(rarities: list[str]) -> list[str]:
    """Return human-readable reasons a set's rarity mix can't be published."""
    total = len(rarities)
    if total == 0:
        return ["Add at least one card."]
    counts = Counter(rarities)
    problems = []

    common_share = counts[COMMON] / total
    if common_share < COMMON_MIN_SHARE:
        problems.append(
            f"At least {COMMON_MIN_SHARE:.0%} of cards must be Common "
            f"(currently {common_share:.0%})."
        )
    for rarity, max_share in RARITY_MAX_SHARE.items():
        share = counts[rarity] / total
        # Allow a single card of any tier so small sets can still have one Legendary.
        if share > max_share and counts[rarity] > 1:
            problems.append(
                f"No more than {max_share:.0%} of cards can be {rarity.capitalize()} "
                f"(currently {share:.0%})."
            )
    return problems
